File: Data Structures & Algorithms/validate-parentheses/submission-3.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for _, c in enumerate(s):
            print(c)
            match c:
                case "(" | "{" | "[":
                    stack.append(c)
                case ")":
                    if not stack or stack.pop() != "(":
                        return False
                case "}":
                    if not stack or stack.pop() != "{":
                        return False
                case "]":
                    if not stack or stack.pop() != "[":
                        return False
                case "_":
                    return False
        return not stack    

File: Data Structures & Algorithms/subsets/submission-3.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> soln = new ArrayList<>();
        Stack<Integer> stack = new Stack<>();

        generateSubsets(nums, 0, stack, soln);
        return soln;
    }

    public void generateSubsets(int[] nums, int i, Stack<Integer> stack, List<List<Integer>> soln) {
        if (nums.length == i) {
            soln.add(new ArrayList<>(stack));
            return;
        }
        
        stack.add(nums[i]);
        generateSubsets(nums, i + 1, stack, soln);
        stack.pop();
        generateSubsets(nums, i + 1, stack, soln);
    }
}
